core/shared/src/main/scala/scalaz/zio/internal/tracing/ZIOFn.scala

```diff
@@ -0,0 +1,35 @@
+package scalaz.zio.internal.tracing
+
+import scalaz.zio.{ UIO, ZIO }
+
+/**
+ * Marks a ZIO utility function that wraps a user-supplied lambda.
+ *
+ * Allows access to the underlying lambda so that stacktraces can
+ * point to supplied user's code, instead of ZIO internals.
+ *
+ * NOTE: it being an `abstract class`, not trait is important as type casing
+ * on class appears to be about 10 times faster in [[scalaz.zio.internal.FiberContext.unwrap]]
+ * hot-spot.
+ * */
+private[zio] abstract class ZIOFn extends Serializable {
```

Comment:
On heap allocations, only the second approach will truly result in less allocations, as the first approach with .traced would have to allocate a new ZIO op. On the first approach, not sure how to replace nested wrappers with it (e.g. in bracket* family). It's definitely worth a try to replace ZIOFn scheme, although I think to postpone it from the current PR as its non-trivial, and a fix for it would not break bincompat.

Reply:
I will definitely try to implement second scheme today, if not succesful - postpone

Reply:
I've replaced some `ZIOFn` uses with a nullable field in `FlatMap` in [another branch](https://github.com/7mind/scalaz-zio/compare/feature/zio-stacktraces...feature/remove-ziofn) - however, I'm not seeing any performance improvement from less allocations in the benchmarks

Reply:
As such, punting this for now

Reply:
Continued here - https://github.com/scalaz/scalaz-zio/issues/916